contracts/Turnstone.vy

```diff
@@ -0,0 +1,105 @@
+# @version 0.3.3
+"""
+@title Turnstone
+@author Volume.Finance
+"""
+
+MAX_VALIDATORS: constant(uint256) = 128
+MAX_BATCH: constant(uint256) = 128
+POWER_THRESHOLD: constant(uint256) = 2_863_311_530 # 2/3 of 2^32
+TURNSTONE_ID: immutable(bytes32)
+
+struct ValsetArgs:
+    validators: DynArray[address, MAX_VALIDATORS]
+    powers: DynArray[uint256, MAX_VALIDATORS]
+    nonce: uint256
+
+struct LogicCallArgs:
+    logicContractAddress: address
+    payload: Bytes[1024]
+
+struct Signature:
+    v: uint256
+    r: uint256
+    s: uint256
+
+event UpdateCheckpoint:
+    new_checkpoint: bytes32
+
+last_checkpoint: public(bytes32)
+last_valset_nonce: public(uint256)
+
+@external
+def __init__(turnstone_id: bytes32, validators: DynArray[address, MAX_VALIDATORS], powers: DynArray[uint256, MAX_VALIDATORS]):
+    TURNSTONE_ID = turnstone_id
+    cumulative_power: uint256 = 0
+    i: uint256 = 0
+    for validator in validators:
+        cumulative_power += powers[i]
+        if cumulative_power >= POWER_THRESHOLD:
+            break
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+    valset_args: ValsetArgs = ValsetArgs({validators: validators, powers: powers, nonce: 0})
+    self.last_valset_nonce = 0
+    new_checkpoint: bytes32 = keccak256(_abi_encode(valset_args.validators, valset_args.powers, method_id=method_id("checkpoint(address[],uint256[])")))
+    self.last_checkpoint = new_checkpoint
+
+@external
+@pure
+def turnstone_id() -> bytes32:
+    return TURNSTONE_ID
+
+@internal
+@pure
+def verify_signature(signer: address, hash: bytes32, sig: Signature) -> bool:
+    message_digest: bytes32 = keccak256(concat(convert("\x19EthereumSignedMessage:\n32", Bytes[26]), hash))
```

Comment:
Side note, if we are spending a lot of gas hashing it might be worth switching to blake2.

Reply:
[If I understood correctly, Blake2 is not perfect for Ethereum](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-152.md)

Reply:
That's interesting! Thank you for the source.